hikari/impl/special_endpoints.py

```diff
@@ -1261,12 +1261,18 @@ def add_button(
     @typing.overload
     def add_button(
         self: _ActionRowBuilderT,
-        style: typing.Union[typing.Literal[messages.ButtonStyle.LINK], typing.Literal[5]],
+        style: typing.Literal[messages.ButtonStyle.LINK, 5],
```

Comment:
Do we really want to use Literals at all? It makes it for issues like what @tandemdude pointed out annoying to fix.

Reply:
what about this is annoying to fix

Reply:
Also do you really want add_button to be returning a union in cases where it can easily be inferred from the passed type that it should be returning a specific type from said union

Reply:
For this specific `add_button` case we should probably have a separate function that sets the specific style to 5. Reason for this being the following example:
```py
import hikari
from hikari.impl import special_endpoints

style = hikari.ButtonStyle.LINK

reveal_type(style)

builder = special_endpoints.ActionRowBuilder().add_button(style, "test")

reveal_type(builder)
```

You will get these logs:
```
test.py:6: note: Revealed type is "hikari.messages.ButtonStyle"
test.py:10: note: Revealed type is "Union[hikari.api.special_endpoints.LinkButtonBuilder[hikari.impl.special_endpoints.ActionRowBuilder*], hikari.api.special_endpoints.InteractiveButtonBuilder[hikari.impl.special_endpoints.ActionRowBuilder*]]"

```
Which still wouldnt fix the issue in hand.

Maybe it would be worth revisiting how how we currently do component builder and switch to a more "embed like" pattern?

Reply:
> Which still wouldnt fix the issue in hand. 

How does this not fix the problem of you not being able to pass a more relaxed type to this function? Also what you just described wouldn't fix the issue as you'd have these functions

```py
    def add_interactive_button(
        self: _ActionRowBuilderT, style: messages.InteractiveButtonTypesT, custom_id: str, /
    ) -> special_endpoints.InteractiveButtonBuilder[_ActionRowBuilderT]:
        ...

    def add_url_button(
        self: _ActionRowBuilderT,
        style: typing.Literal[messages.ButtonStyle.LINK, 5],
        url: str,
        /,
    ) -> special_endpoints.LinkButtonBuilder[_ActionRowBuilderT]:
        ...

```

Neither of which you could pass something typed as `hikari.ButtonStyle` to, also what you just showed in your example is literally intended cause you passed something typed as hikari.ButtonStyle rather than a literal type.

Reply:
> switch to a more "embed like" pattern?

Ok so you know how you don't need an emoji object here, just an id or name, how would you deal with merging that with a data model class where you get a full emoji object without just making it less relaxed for no good reason.

And another question, isn't the embed approach just kinda bad but left there cause its too late to switch to a proper strict builder/dto pattern where's there's like actual separation. Like that pattern's already inferring that you can just safely relay received embeds which you can't cause there's like a ton of received fields you cannot send and it also breaks away from the lib standard that DTO models are considered immutable and probably also messes with the cache's reliability in the process because of that.